singularity-eos/eos/eos_eospac.hpp

```diff
@@ -72,103 +72,1156 @@ class EOSPAC : public EosBase<EOSPAC> {
+          eosSafeInterpolate(&table, nxypairs, R.data(), P.data(), T.data(), dx, dy,
+                             "TofRP", Verbosity::Quiet);
+        } else {
+          UNDEFINED_ERROR;
+        }
+
+        for (int i = 0; i < nelements; ++i) {
+          int j = istart + i;
+          temps[j] = temperatureFromSesame(T[i]);
+        }
+      }
+      if ((output & thermalqs::specific_internal_energy) ||
+          (output & thermalqs::specific_heat || output & thermalqs::bulk_modulus)) {
+        EOS_INTEGER table = EofRT_table_;
+        eosSafeInterpolate(&table, nxypairs, R.data(), T.data(), E.data(), DEDR.data(),
+                           DEDT.data(), "EofRT", Verbosity::Quiet);
+      }
+      if ((output & thermalqs::pressure) || (output & thermalqs::bulk_modulus)) {
+        EOS_INTEGER table = PofRT_table_;
+        eosSafeInterpolate(&table, nxypairs, R.data(), T.data(), P.data(), DPDR.data(),
+                           DPDT.data(), "PofRT", Verbosity::Quiet);
+        // Thermodynamics: Bt = rho*dP/drho|T, Bs=Bt*Cv/Cp, Cv=dE/dT|v, and
+        // Cp-Cv=-T(dP/dT|v)^2/(dP/dV|T) or Cp-Cv=-T/rho^2 (dP/dT|v)^2/(dP/drho|T)
+        // Therefore: Bs=Bt Cv/(CV+above)
+        // BMOD = rho*dx[0]*CV/(CV-T[0]/(rho*rho)*dy[0]*dy[0]/dx[0]);
+      }
+      if (output & thermalqs::specific_internal_energy) {
+        for (int i = 0; i < nelements; ++i) {
+          int j = istart + i;
+          energies[j] = sieFromSesame(E[i]);
+        }
+      }
+      if (output & thermalqs::pressure) {
+        for (int i = 0; i < nelements; ++i) {
+          int j = istart + i;
+          presses[j] = pressureFromSesame(P[i]);
+        }
+      }
+      if (output & thermalqs::specific_heat) {
+        for (int i = 0; i < nelements; ++i) {
+          int j = istart + i;
+          cvs[j] =
+              cvFromSesame(std::max(DEDT[i], 0.0)); // Here we do something to the data!
+        }
+      }
+      if (output & thermalqs::bulk_modulus) {
+        for (int i = 0; i < nelements; ++i) {
+          int j = istart + i;
+          const Real rho = R[i];
+          Real BMOD = 0.0;
+          if (DEDT[i] > 0.0 && rho > 0.0) {
+            const Real DPDE = DPDT[i] / DEDT[i];
+            BMOD = rho * DPDR[i] + DPDE * (P[i] / rho - rho * DEDR[i]);
+          } else if (rho > 0.0) { // Case: DEDT <= 0
+            // We need a different DPDE call apparently????
+            // In xRAGE, they call out to P(rho,e) in this case to get the
+            // derivative directly from the half-inverted table.
+            // But upon further review,
+            // I think it will end up evaluating to BMOD_T in any case because cv will
+            // be zero! See eos_eospac.f90 line 1261 BMOD =
+            // BMOD_T+DPDE*DPDE*std::max(DEDT,0.0)*T[0]/rho;
+            // BMOD_T = std::max(rho * DPDR[i], 0.0);
+            // BMOD = BMOD_T; //+DPDE*DPDE*std::max(DEDT,0.0)*T[0]/rho;
+            BMOD = std::max(rho * DPDR[i], 0.0);
```

Comment:
Honestly there's not much we can do when the heat capacity (DEDT) is zero. If it's negative then the EOS is fundamentally flawed.

I'd be in favor of an assert here so that if somebody gets crazy results they can run in debug mode and see that their EOS has serious issues, but since this code was here before your modifications I'll create an issue for this so we can handle it later.

Reply:
I think this probably merits a larger discussion.

Reply:
Discussion: https://github.com/lanl/singularity-eos/issues/180